File: codigo/02 - Regresion/3 - Polinomica/polinomic.py
# -*- coding: utf-8 -*-
"""
Created on Fri Dec 18 12:02:08 2020

@author: msantamaria
"""

# Regresión polinómica

# Cómo importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el data set 
dataset = pd.read_csv("Position_Salaries.csv")
X = dataset.iloc[:,1:2].values # iloc -> index localization
y = dataset.iloc[:,2].values

# Dividir el dataset en conjunto de entrenamiento y en conjunto de testing
# Tenemos muy pocos datos como para dividirlos
 
# Ajustar la regresión lineal con el dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Ajustar la regresión polinómica con todo el dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree=4)
X_poly = poly_reg.fit_transform(X) 
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly,y)

# Visualización de los resultados del Modelo Lineal
plt.scatter(X,y,color='red')
plt.plot(X,lin_reg.predict(X),color='blue')
plt.title("Modelo de Regresión Lineal")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (En dolares)")
plt.show()

# Visualización de los resultados del Modelo Polinómico
X_grid = np.arange(min(X),max(X),0.1)
X_grid = X_grid.reshape(len(X_grid),1)
plt.scatter(X,y,color='red')
plt.plot(X_grid,lin_reg2.predict(poly_reg.fit_transform(X_grid)),color='blue')
plt.title("Modelo de Regresión Polinómica")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (En dolares)")
plt.show()

# Predicción de nuestros modelos
lin_reg.predict([[6.5]])
lin_reg2.predict(poly_reg.fit_transform([[6.5]]))

File: codigo/02 - Regresion/3 - Polinomica/polinomic.R
# Regresión Polinómica

# Importar el data set
dataset = read.csv("2 - Regresion/3 - Polinomica/Position_Salaries.csv")
dataset = dataset[,2:3]

# Ajustar modelo de regresión lineal con el conjunto de datos
lin_reg = lm(formula = Salary ~ ., data = dataset)

# Ajustar Modelo de Regresión Polinómica con el Conjunto de Datos
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ ., data = dataset)

# Visualización del modelo lineal
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary), color = "red") +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)), color = "blue") + 
  ggtitle("Predicción lineal del sueldo en función del nivel del empleado") + 
  xlab("Nivel del empleado") + 
  ylab("Sueldo (en dolares)")

# Visualización del modelo polinómico
x_grid = seq(min(dataset$Level),max(dataset$Level),0.1)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary), color = "red") +
  geom_line(aes(x = x_grid, y = predict(poly_reg, newdata = data.frame(Level = x_grid,
                                                                       Level2 = x_grid^2,
                                                                       Level3 = x_grid^3,
                                                                       Level4 = x_grid^4))), 
            color = "blue") + 
  ggtitle("Predicción lineal del sueldo en función del nivel del empleado") + 
  xlab("Nivel del empleado") + 
  ylab("Sueldo (en dolares)")

# Predicción de nuevos resultados con Regresión Lineal
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))

# Predicción de nuevos resultados con Regresión Polinómica
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))
